### PR TITLE
Draft: Adding new worker which uses PSI/J to run tasks

### PR DESCRIPTION
**Pull request by adi611**
## Types of changes

- New feature (non-breaking change which adds functionality)

## Summary

Add a new worker called `PsijWorker` to `workers.py` which uses PSI/J to run tasks.

## Checklist

- [ ] I have added tests to cover my changes (if necessary)
- [ ] I have updated documentation (if necessary)

### CODE REVIEW

**Comment by djarecka** [comment]
i'm trying to understand why the slurm test fails now (last run I cancelled, but it took 5h...), since you didn't any new test here yet. any idea?

**Comment by djarecka** [comment]
@adi611 - perhaps you can merge current master and try new `actions/checkout` that was introduced.  I've run `test_shelltask.py` with the docker image on my laptop and it was fine

**Comment by adi611** [comment]
> @adi611 - perhaps you can merge current master and try new `actions/checkout` that was introduced. I've run `test_shelltask.py` with the docker image on my laptop and it was fine

Ok I will try this out

**Comment by adi611** [comment]
> i'm trying to understand why the slurm test fails now (last run I cancelled, but it took 5h...), since you didn't any new test here yet. any idea?

I couldn't find any reason for the workflow to not work here, when it works elsewhere. I will try the new `actions/checkout` and check if the problem persists.

**Comment by adi611** [comment]
I have updated the worker. Currently I am using the `slurm` instance for `psij.JobExecutor` to run Slurm tests.

Prerequisite:
- The local pydra repository must be mounted as `/pydra` as done in the Slurm GA workflow
- Install `psij-python` in the docker container
- Change the plugin for the test to `psij` or make changes in conftest file as discussed in #691 

Features:
- Can run the Slurm tests individually 

Known issues:
- Tests fail when running many simultaneously. For e.g., `pydra/pydra/engine/tests/test_shelltask.py::test_shell_cmd_4` pass but `test_shell_cmd_4` fails when testing all tests in `test_shelltask.py`.
- Not clean code

I am working on fixing these issues.

**Comment by djarecka** [comment]
@adi611 - I've had problems with my MIT account during the weekend, so I was not able to test it, but hopefully will be able to solve my issues tomorrow

**Comment by adi611** [comment]
@djarecka - I think the issue arises when running the tests with the `psij` plugin and `-n auto` mode of `pytest`. I tried running all the tests in `test_shelltask.py`:
- Using the `local` instance of the `psij` job executor, with the command `pytest --color=yes -v pydra/engine/tests/test_shelltask.py`. Here is the result: (2 tests fail, others run and pass without any issue)
```
========================================== short test summary info ==========================================
FAILED pydra/engine/tests/test_shelltask.py::test_shell_cmd_outputspec_1b_exception[psij] - Failed: DID NOT RAISE <class 'Exception'>
FAILED pydra/engine/tests/test_shelltask.py::test_shell_cmd_outputspec_2a_exception[psij] - Failed: DID NOT RAISE <class 'Exception'>
====================== 2 failed, 147 passed, 1 skipped, 3 xfailed in 388.45s (0:06:28) ======================
```
- Using the Slurm container and the `slurm` instance of `psij` job executor (without `-n auto`): the tests run and pass, but at a very slow pace
- Using the Slurm container and the `slurm` instance of `psij` job executor (with `-n auto`): many tests fail

**Comment by adi611** [comment]
I have fixed the issue with pytest's `-n auto` mode; it can now run multiple tests at once.

**Comment by djarecka** [comment]
@adi611 - I've tried to run the tests with new worker on our cluster, but tests are failing, I'm not able to get results when running with `Submitter`. Will try to check debug it in the next few days

**Comment by adi611** [comment]
> @adi611 - I've tried to run the tests with new worker on our cluster, but tests are failing, I'm not able to get results when running with `Submitter`. Will try to check debug it in the next few days

I think the issue might be with the paths to files and functions (like `/pydra/pydra/engine/run_pickled_function.py`), and/or the `python3.9` command. I used the previous Slurm container as reference for writing this and I should make it more generalized now. Could you please check if the issue still persists after correcting the paths and python command? Meanwhile I will update the PR.

**Comment by djarecka** [comment]
yes, it should be more generalized, you can always use the task `output_dir`

**Comment by adi611** [comment]
I think it should work now

**Comment by adi611** [comment]
Added the option to switch between the different schedulers provided by `psij`. For e.g., `plugin='psij-local'` and `plugin='psij-slurm'` will use `local` and `slurm` instances for `psij`'s job executor respectively. For now, the list contains only `local` and `slurm` options.

**Comment by djarecka** [comment]
could you please modify `conftest` so we can start testing the new workers in GA. 

I've started testing `psij-local` on my osx and  5 tests are failing. Could you try to debug them? 

I also run a subset of tests on slurm on the MIT cluster and it looks like they are fixed! :) but I will test more tomorrow!

**Comment by adi611** [comment]
> could you please modify `conftest` so we can start testing the new workers in GA.

I'm currently working on it.
> I've started testing `psij-local` on my osx and 5 tests are failing. Could you try to debug them?

 Can you please tell me which tests are failing?

> I also run a subset of tests on slurm on the MIT cluster and it looks like they are fixed! :) but I will test more tomorrow!

Great!

**Comment by adi611** [comment]
The `psij-local` and `psij-slurm` pass all the required tests, though `psij-slurm` takes much longer than the actual `slurm` plugin. I will update the `conftest.py` file to resolve the conflicts with other tests. Please let me know if I need to make other changes.

**Comment by adi611** [comment]
I will also try to improve the performance of `psij-slurm`

**Comment by djarecka** [comment]
with new version the `psij-local` works on my osx! will check the new version of slurm tomorrow morning.

for the conftest, you could just expand the list of worker and add `psij-local` to `cf` and `psij-slurm` to `slurm` and see how long it takes on GA

**Comment by adi611** [comment]
> for the conftest, you could just expand the list of worker and add `psij-local` to `cf` and `psij-slurm` to `slurm` and see how long it takes on GA

I have updated the `conftest.py` to add the `psij` plugin options. The fixtures `--psij=local` and `--psij=slurm` correspond to the `psij-local` and `psij-slurm` plugins respectively.  Also, I have created specific GA workflows for the `psij` plugins: `PSI/J-Local` and `PSI/J-SLURM`.

**Comment by adi611** [comment]
For `PSI/J-SLURM` it took 1h 56m 35s to complete, as can be seen [here](https://github.com/nipype/pydra/actions/runs/6254101591/job/16980933062?pr=694).

File: pydra/engine/workers.py [hunk 1]
```diff
@@ -1,901 +1,1043 @@
...
+        if isinstance(runnable, TaskBase):
+            cache_dir = runnable.cache_dir
+            file_path = os.path.join(cache_dir, "my_function.pkl")
+            with open(file_path, "wb") as file:
+                pickle.dump(runnable._run, file)
+            func_path = os.path.join(absolute_path, "run_pickled.py")
+            spec = self.make_spec("python", [func_path, file_path])
+        else:  # it could be tuple that includes pickle files with tasks and inputs
+            cache_dir = runnable[-1].cache_dir
+            file_path_1 = os.path.join(cache_dir, "my_function.pkl")
```
Comment (djarecka) [inline pydra/engine/workers.py:988]: it's ok to use `os.path`, but we mostly use `Path` from `pathlib` and build new path [this way](https://github.com/nipype/pydra/blob/0245cdcf940abcdf58443f754e958a0b1bcb02c5/pydra/engine/core.py#L414)

**Review by djarecka — COMMENTED** [review]
I left some small suggestions, but I also think we should try to run `psij-local` with docker. It seems like all `test_docker` are disabled when `PSI/J-Local / test `

File: pydra/engine/workers.py [hunk 2]
```diff
@@ -1,901 +1,1043 @@
...
-
-
-WORKERS = {
-    "serial": SerialWorker,
-    "cf": ConcurrentFuturesWorker,
-    "slurm": SlurmWorker,
-    "dask": DaskWorker,
-    "sge": SGEWorker,
-}
+"""Execution workers."""
```
[the remark this answers is not in the document]
Comment (adi611) [reply pydra/engine/workers.py:1]: Maybe something to do with CRLF and LF. I will try to correct it.

File: pydra/engine/workers.py [hunk 3]
```diff
@@ -1,901 +1,1043 @@
...
+                stderr_contents = stderr_file.read()
+            raise Exception(
+                f"stderr_path '{spec.stderr_path}' is not empty. Contents:\n{stderr_contents}"
+            )
+
+        return
+
+    def close(self):
+        """Finalize the internal pool of tasks."""
+        pass
```
[the remark this answers is not in the document]
Comment (adi611) [reply pydra/engine/workers.py]: I have gone through the PSIJ workflow examples once again, but I am unable to find anything similar to this.

File: pydra/engine/workers.py [hunk 4]
```diff
@@ -1,901 +1,1043 @@
...
+            Scheduler for PSI/J.
+        """
+        try:
+            import psij
+        except ImportError:
+            logger.critical("Please install psij.")
+            raise
+        logger.debug("Initialize PsijWorker")
+        self.psij = psij
+        self.subtype = subtype
```
[the remark this answers is not in the document]
Comment (adi611) [reply pydra/engine/workers.py]: Added

File: pydra/engine/workers.py [hunk 5]
```diff
@@ -1,901 +1,1043 @@
...
+        spec.stderr_path = os.path.join(cache_dir, "demo.stderr")
+
+        job = self.make_job(spec, None)
+        jex.submit(job)
+        job.wait()
+
+        if os.path.getsize(spec.stderr_path) > 0:
+            with open(spec.stderr_path, "r") as stderr_file:
+                stderr_contents = stderr_file.read()
+            raise Exception(
```
[the remark this answers is not in the document]
Comment (adi611) [reply pydra/engine/workers.py]: I think we already check for exceptions in `test_shell_cmd_outputspec_1b_exception` and `test_shell_cmd_outputspec_2a_exception` in `test_shelltask`. Should I add something similar to `test_submitter`?

File: pydra/engine/workers.py (hunk 1 above)
Comment (adi611) [reply pydra/engine/workers.py]: How can I get the absolute location of `run_pickled.py` using this?

**Comment by adi611** [comment]
@djarecka - I checked the `test_dockertask` tests in `PSI/J-Local / test ` and all the tests which pass for `cf` pass for `psij-local`, for e.g., `test_docker_1` pass while `test_docker_3` gets skipped, for both `cf` and `psij-local`.

File: pydra/engine/workers.py (hunk 1 above)
Comment (djarecka) [reply pydra/engine/workers.py]: check https://docs.python.org/3/library/pathlib.html#pathlib.Path.resolve

File: pydra/engine/workers.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (djarecka) [reply pydra/engine/workers.py]: you can actually reuse this tests! don't have to be in `test_submitter`

[the remark this answers is not in the document]
Comment (djarecka) [reply pydra/engine/workers.py]: ok, you're already running it

File: pydra/engine/workers.py (hunk 1 above)
Comment (adi611) [reply pydra/engine/workers.py]: Done

File: pydra/engine/workers.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (djarecka) [reply pydra/engine/workers.py]: ok, thank you for checking

**Pull request merged**